File: web/backend/services/docker.js
// 封装 Dockerode 服务，负责管理出口 Egress 容器的生命周期
// 中文注释，保障逻辑清晰及错误处理

const Docker = require('dockerode');

// 初始化 Docker 客户端，读取挂载的 docker.sock
const docker = new Docker({ socketPath: '/var/run/docker.sock' });

// 镜像名称
const EGRESS_IMAGE = 'vpngate-egress:latest';

// 缓存宿主机数据目录路径，避免每次启动容器都执行 inspect
let cachedHostDataPath = null;

/**
 * 推导宿主机上 /app/data 挂载的真实绝对路径
 * 优先使用环境变量 HOST_DATA_PATH，否则通过 inspect 面板自身容器获取
 */
async function resolveHostDataPath() {
  // 如果已有缓存，直接返回
  if (cachedHostDataPath) return cachedHostDataPath;

  // 方式一：通过环境变量显式指定（docker-compose 中配置）
  if (process.env.HOST_DATA_PATH) {
    cachedHostDataPath = process.env.HOST_DATA_PATH;
    console.log(`[+] 从环境变量获取宿主机 Data 路径: ${cachedHostDataPath}`);
    return cachedHostDataPath;
  }

  // 方式二：通过 Docker inspect 面板自身容器，自动探测绑定挂载源路径
  try {
    const selfContainer = docker.getContainer('vless-web-panel');
    const selfInfo = await selfContainer.inspect();
    const dataMount = selfInfo.Mounts.find(m => m.Destination === '/app/data');
    if (dataMount && dataMount.Source) {
      cachedHostDataPath = dataMount.Source;
      console.log(`[+] 自动探测到宿主机 Data 真实路径: ${cachedHostDataPath}`);
      return cachedHostDataPath;
    }
  } catch (err) {
    console.warn('[!] 自动探测宿主机挂载路径失败:', err.message);
  }

  // 方式三：回退方案
  cachedHostDataPath = '/app/data';
  console.warn('[!] 未能推导宿主机路径，使用容器内路径 /app/data（在 docker.sock 模式下可能导致挂载失败）');
  return cachedHostDataPath;
}

/**
 * 启动或创建出口容器
 * @param {Object} egress - 出口对象
 * @param {string} egress.name - 出口名称
 * @param {number} egress.port - 宿主机暴露端口
 */
async function startEgressContainer(egress) {
  const containerName = `egress-${egress.name}`;

  const hostDataPath = await resolveHostDataPath();
  const hostEgressDir = `${hostDataPath}/egress/${egress.name}`;

  // 1. 检查同名容器是否已存在，如果存在则先停止并删除
  try {
    const existingContainer = docker.getContainer(containerName);
    const info = await existingContainer.inspect();
    console.log(`[*] 检测到重名容器 ${containerName} (状态: ${info.State.Status})，正在进行清理...`);
    if (info.State.Running) {
      await existingContainer.stop({ t: 5 });
    }
    await existingContainer.remove();
    console.log(`[+] 旧容器 ${containerName} 清理完毕`);
  } catch (error) {
    // 容器不存在，忽略
  }

  // 2. 创建并启动新容器
  console.log(`[*] 正在创建出口容器 ${containerName}，宿主机映射端口: ${egress.port}...`);
  
  const container = await docker.createContainer({
    Image: EGRESS_IMAGE,
    name: containerName,
    HostConfig: {
      PortBindings: {
        '10000/tcp': [{ HostPort: String(egress.port) }]
      },
      CapAdd: ['NET_ADMIN'], // OpenVPN 需要网络管理权限
      Devices: [
        {
          PathOnHost: '/dev/net/tun',
          PathInContainer: '/dev/net/tun',
          CgroupPermissions: 'rwm'
        }
      ],
      Binds: [
        `${hostEgressDir}/client.ovpn:/etc/openvpn/client.ovpn:ro`,
        `${hostEgressDir}/config.json:/etc/xray/config.json:ro`
      ],
      // 由 self-healing.js 统一决策是否漂移，容器只需保证始终可重启。
      // 使用 unless-stopped 策略避免 on-failure 重试上限耗尽后容器永久停止。
      RestartPolicy: { Name: 'unless-stopped' },
      // 将出口容器加入面板同一网络，便于后续内部通信扩展
      NetworkMode: 'vless-net'
    }
  });

  await container.start();
  console.log(`[+] 出口容器 ${containerName} 启动成功，容器ID: ${container.id}`);
  return container.id;
}

/**
 * 停止并删除出口容器
 * @param {string} name - 出口名称
 */
async function stopAndRemoveContainer(name) {
  const containerName = `egress-${name}`;
  try {
    const container = docker.getContainer(containerName);
    const info = await container.inspect();
    
    console.log(`[*] 正在停止出口容器 ${containerName}...`);
    if (info.State.Running) {
      await container.stop({ t: 5 });
    }
    
    console.log(`[*] 正在删除出口容器 ${containerName}...`);
    await container.remove();
    console.log(`[+] 出口容器 ${containerName} 删除成功`);
    return true;
  } catch (error) {
    console.log(`[*] 容器 ${containerName} 未运行或不存在，无需删除`);
    return false;
  }
}

/**
 * 检查容器状态并获取当前 VPN 出口的真实 IP
 * 使用 Dockerode 原生的 demuxStream 正确解析 Docker 多路复用 stream 协议头
 * @param {string} name - 出口名称
 */
async function getContainerStatusAndIp(name) {
  const containerName = `egress-${name}`;
  try {
    const container = docker.getContainer(containerName);
    const info = await container.inspect();
    const isRunning = info.State.Running;
    const status = info.State.Status; // running, exited, etc.

    if (!isRunning) {
      return { status, ip: 'offline', error: '容器未运行' };
    }

    // 简体中文注释：采用多源容错与随机打散，防止单点故障及 429 速率限制
    const ipApis = [
      'https://api.ipify.org',
      'https://ifconfig.me/ip',
      'https://ipinfo.io/ip'
    ];
    const shuffledApis = [...ipApis].sort(() => Math.random() - 0.5);
    const cmdString = shuffledApis.map(api => `curl -fs --interface tun0 --max-time 4 ${api}`).join(' || ');

    // 在容器内部执行 curl 获取 VPN 出口真实 IP，支持多源回退与防限流
    try {
      const exec = await container.exec({
        Cmd: ['sh', '-c', cmdString],
        AttachStdout: true,
        AttachStderr: true
      });

      const stream = await exec.start({ Detach: false });
      
      return new Promise((resolve) => {
        let stdout = '';
        let stderr = '';

        // 使用 Dockerode 官方提供的 demuxStream 工具方法
        // 正确分离 stdout/stderr 的多路复用协议帧（每帧前 8 字节为协议头）
        const { PassThrough } = require('stream');
        const stdoutStream = new PassThrough();
        const stderrStream = new PassThrough();

        docker.modem.demuxStream(stream, stdoutStream, stderrStream);

        stdoutStream.on('data', (chunk) => {
          stdout += chunk.toString();
        });
        stderrStream.on('data', (chunk) => {
          stderr += chunk.toString();
        });

        // 设置超时保护，避免 exec 挂起导致健康检查阻塞
        const timeout = setTimeout(() => {
          resolve({ status: 'running', ip: 'error', error: '容器内 IP 检测超时' });
        }, 10000);

        stream.on('end', () => {
          clearTimeout(timeout);
          const cleanIp = stdout.trim();
          
          // 验证 IP 格式是否为标准 IPv4
          if (cleanIp && /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(cleanIp)) {
            resolve({ status: 'running', ip: cleanIp });
          } else {
            resolve({ status: 'error', ip: 'error', error: `无法获取有效出口 IP，stdout: ${cleanIp}, stderr: ${stderr.trim()}` });
          }
        });

        stream.on('error', (err) => {
          clearTimeout(timeout);
          resolve({ status: 'error', ip: 'error', error: err.message });
        });
      });
    } catch (execErr) {
      return { status: 'running', ip: 'error', error: `执行 IP 检测失败: ${execErr.message}` };
    }
  } catch (error) {
    return { status: 'offline', ip: 'offline', error: '容器未创建或已被删除' };
  }
}

/**
 * 获取出口容器最近日志，便于面板诊断 OpenVPN / Xray 启动失败原因
 * @param {string} name - 出口名称
 * @param {number} tail - 返回最近多少行日志
 */
async function getContainerLogs(name, tail = 120) {
  const containerName = `egress-${name}`;
  try {
    const container = docker.getContainer(containerName);
    await container.inspect();

    const logBuffer = await container.logs({
      stdout: true,
      stderr: true,
      timestamps: true,
      tail
    });

    // docker logs 在非 TTY 容器下会带 8 字节 stream header，这里做一次轻量清理。
    let offset = 0;
    const chunks = [];
    while (offset + 8 <= logBuffer.length) {
      const frameSize = logBuffer.readUInt32BE(offset + 4);
      const frameStart = offset + 8;
      const frameEnd = frameStart + frameSize;
      if (frameEnd > logBuffer.length) break;
      chunks.push(logBuffer.subarray(frameStart, frameEnd));
      offset = frameEnd;
    }

    if (chunks.length > 0) {
      return Buffer.concat(chunks).toString('utf-8');
    }
    return logBuffer.toString('utf-8');
  } catch (error) {
    throw new Error(`读取容器日志失败: ${error.message}`);
  }
}

module.exports = {
  startEgressContainer,
  stopAndRemoveContainer,
  getContainerStatusAndIp,
  getContainerLogs
};

File: web/backend-go/services/docker.go
// 封装 Docker SDK，负责管理出口 Egress 容器的生命周期与测活
// 简体中文注释，基于官方 Docker SDK 实现，支持容器快速重启复用与 demux 拆流解复用

package services

import (
	"bytes"
	"context"
	"fmt"
	"math/rand"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"vless-reality-panel/models"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"
	"github.com/docker/docker/pkg/stdcopy"
	"github.com/docker/go-connections/nat"
)

const (
	egressImage     = "vpngate-egress:latest"
	cacheMaxAge     = 6 * time.Second
	containerPrefix = "egress-"
)

// ContainerStatusReport 容器测活状态报告
type ContainerStatusReport struct {
	Status string `json:"status"`
	IP     string `json:"ip"`
	Error  string `json:"error"`
}

var (
	dockerCli           *client.Client
	cachedHostDataPath  string
	hostPathOnce        sync.Once
	statusCache         = make(map[string]cachedReport)
	statusCacheMutex    sync.RWMutex
	ipv4Regexp          = regexp.MustCompile(`^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$`)
	initDockerClientOnce sync.Once
	dockerInitErr       error
)

type cachedReport struct {
	timestamp time.Time
	data      *ContainerStatusReport
}

// 获取/初始化 Docker Client 单例
func getDockerClient() (*client.Client, error) {
	initDockerClientOnce.Do(func() {
		dockerCli, dockerInitErr = client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
		if dockerInitErr != nil {
			dockerInitErr = fmt.Errorf("初始化 Docker 客户端失败: %v", dockerInitErr)
		}
	})
	return dockerCli, dockerInitErr
}

// resolveHostDataPath 自动推导宿主机挂载的真实数据目录
func resolveHostDataPath() string {
	hostPathOnce.Do(func() {
		// 方式一：环境变量指定
		if envPath := os.Getenv("HOST_DATA_PATH"); envPath != "" {
			cachedHostDataPath = envPath
			fmt.Printf("[*] 从环境变量获取宿主机 Data 路径: %s\n", cachedHostDataPath)
			return
		}

		// 方式二：Inspect 面板自身容器
		cli, err := getDockerClient()
		if err == nil {
			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
			defer cancel()
			if info, err := cli.ContainerInspect(ctx, "vless-web-panel"); err == nil {
				for _, m := range info.Mounts {
					if m.Destination == "/app/data" {
						cachedHostDataPath = m.Source
						fmt.Printf("[*] 自动探测到宿主机 Data 真实路径: %s\n", cachedHostDataPath)
						return
					}
				}
			}
		}

		// 方式三：默认兜底
		cachedHostDataPath = "/app/data"
		fmt.Printf("[!] 未能推导宿主机路径，使用默认值: %s\n", cachedHostDataPath)
	})
	return cachedHostDataPath
}

// StartEgressContainer 启动或创建出口容器，若存在则快速重启复用
func StartEgressContainer(egress *models.Egress) (string, error) {
	cli, err := getDockerClient()
	if err != nil {
		return "", err
	}

	containerName := containerPrefix + egress.Name
	hostDataPath := resolveHostDataPath()
	hostEgressDir := fmt.Sprintf("%s/egress/%s", hostDataPath, egress.Name)

	ctx := context.Background()

	// 1. 检查同名容器是否已存在，如果存在则直接重启复用
	info, err := cli.ContainerInspect(ctx, containerName)
	if err == nil {
		// 容器存在，执行快速重启复用
		fmt.Printf("[*] 检测到重名容器 %s (状态: %s)，执行快速重启复用...\n", containerName, info.State.Status)
		stopTimeout := 5 * time.Second
		err = cli.ContainerRestart(ctx, containerName, &stopTimeout)
		if err != nil {
			return "", fmt.Errorf("快速重启容器失败: %v", err)
		}
		fmt.Printf("[+] 出口容器 %s 重启成功\n", containerName)
		return info.ID, nil
	}

	// 2. 容器不存在，执行创建并启动流程
	fmt.Printf("[*] 正在创建出口容器 %s，宿主机映射端口: %d...\n", containerName, egress.Port)

	portBind := nat.PortMap{
		"10000/tcp": []nat.PortBinding{
			{HostPort: strconv.Itoa(egress.Port)},
		},
	}

	resp, err := cli.ContainerCreate(ctx,
		&container.Config{
			Image: egressImage,
			ExposedPorts: nat.PortSet{
				"10000/tcp": struct{}{},
			},
		},
		&container.HostConfig{
			PortBindings: portBind,
			CapAdd:       []string{"NET_ADMIN"}, // OpenVPN 需要网络管理权限
			Resources: container.Resources{
				Devices: []container.DeviceMapping{
					{
						PathOnHost:        "/dev/net/tun",
						PathInContainer:   "/dev/net/tun",
						CgroupPermissions: "rwm",
					},
				},
			},
			Binds: []string{
				fmt.Sprintf("%s/client.ovpn:/etc/openvpn/client.ovpn:ro", hostEgressDir),
				fmt.Sprintf("%s/config.json:/etc/xray/config.json:ro", hostEgressDir),
			},
			RestartPolicy: container.RestartPolicy{
				Name: "unless-stopped",
			},
			NetworkMode: "vless-net",
		},
		nil, nil, containerName,
	)
	if err != nil {
		return "", fmt.Errorf("创建容器失败: %v", err)
	}

	err = cli.ContainerStart(ctx, resp.ID, types.ContainerStartOptions{})
	if err != nil {
		// 启动失败尝试清理残留
		_ = cli.ContainerRemove(ctx, resp.ID, types.ContainerRemoveOptions{Force: true})
		return "", fmt.Errorf("启动容器失败: %v", err)
	}

	fmt.Printf("[+] 出口容器 %s 启动成功，容器ID: %s\n", containerName, resp.ID)
	return resp.ID, nil
}

// StopAndRemoveContainer 停止并删除出口容器
func StopAndRemoveContainer(name string) error {
	cli, err := getDockerClient()
	if err != nil {
		return err
	}

	containerName := containerPrefix + name
	ctx := context.Background()

	info, err := cli.ContainerInspect(ctx, containerName)
	if err != nil {
		if client.IsErrNotFound(err) {
			fmt.Printf("[*] 容器 %s 未创建或已不存在，无需删除\n", containerName)
			return nil
		}
		return err
	}

	fmt.Printf("[*] 正在停止出口容器 %s...\n", containerName)
	if info.State.Running {
		stopTimeout := 5 * time.Second
		err = cli.ContainerStop(ctx, containerName, &stopTimeout)
		if err != nil {
			fmt.Printf("[!] 停止容器 %s 失败: %v，尝试强制删除\n", containerName, err)
		}
	}

	fmt.Printf("[*] 正在删除出口容器 %s...\n", containerName)
	err = cli.ContainerRemove(ctx, containerName, types.ContainerRemoveOptions{Force: true})
	if err != nil {
		return fmt.Errorf("删除容器 %s 失败: %v", containerName, err)
	}

	fmt.Printf("[+] 出口容器 %s 删除成功\n", containerName)
	return nil
}

// GetContainerStatusAndIp 获取状态与 IP (含 6 秒缓存包装)
func GetContainerStatusAndIp(name string) (*ContainerStatusReport, error) {
	statusCacheMutex.RLock()
	cached, found := statusCache[name]
	if found && time.Since(cached.timestamp) < cacheMaxAge {
		defer statusCacheMutex.RUnlock()
		return cached.data, nil
	}
	statusCacheMutex.RUnlock()

	statusCacheMutex.Lock()
	defer statusCacheMutex.Unlock()
	// 双检锁
	if cached, found := statusCache[name]; found && time.Since(cached.timestamp) < cacheMaxAge {
		return cached.data, nil
	}

	report, err := getContainerStatusAndIpRaw(name)
	if err != nil {
		return nil, err
	}

	statusCache[name] = cachedReport{
		timestamp: time.Now(),
		data:      report,
	}
	return report, nil
}

// 容器测活获取外部 IP 的底层实现
func getContainerStatusAndIpRaw(name string) (*ContainerStatusReport, error) {
	cli, err := getDockerClient()
	if err != nil {
		return nil, err
	}

	containerName := containerPrefix + name
	ctx := context.Background()

	info, err := cli.ContainerInspect(ctx, containerName)
	if err != nil {
		if client.IsErrNotFound(err) {
			return &ContainerStatusReport{Status: "offline", IP: "offline", Error: "容器未创建或已被删除"}, nil
		}
		return nil, err
	}

	if !info.State.Running {
		return &ContainerStatusReport{Status: info.State.Status, IP: "offline", Error: "容器未运行"}, nil
	}

	// 简体中文注释：IP 探测源列表随机打散，执行多源回退测活，降低 429 被封禁概率
	ipApis := []string{
		"https://api.ipify.org",
		"https://ifconfig.me/ip",
		"https://ipinfo.io/ip",
	}
	// 随机排序
	rand.Seed(time.Now().UnixNano())
	shuffled := make([]string, len(ipApis))
	copy(shuffled, ipApis)
	rand.Shuffle(len(shuffled), func(i, j int) {
		shuffled[i], shuffled[j] = shuffled[j], shuffled[i]
	})

	var cmdList []string
	for _, api := range shuffled {
		cmdList = append(cmdList, fmt.Sprintf("curl -fs --interface tun0 --max-time 4 %s", api))
	}
	cmdString := strings.Join(cmdList, " || ")

	// 容器内执行 Exec
	execConfig := types.ExecConfig{
		AttachStdout: true,
		AttachStderr: true,
		Cmd:          []string{"sh", "-c", cmdString},
	}

	execCreateResp, err := cli.ContainerExecCreate(ctx, containerName, execConfig)
	if err != nil {
		return &ContainerStatusReport{Status: "running", IP: "error", Error: fmt.Sprintf("创建 Exec 失败: %v", err)}, nil
	}

	// 建立 Exec 连接并获取 ioStream
	resp, err := cli.ContainerExecAttach(ctx, execCreateResp.ID, types.ExecStartCheck{})
	if err != nil {
		return &ContainerStatusReport{Status: "running", IP: "error", Error: fmt.Sprintf("启动 Exec 失败: %v", err)}, nil
	}
	defer resp.Close()

	// 拆流缓冲区
	var stdoutBuf, stderrBuf bytes.Buffer
	doneChan := make(chan error, 1)

	go func() {
		// 简体中文注释：使用 StdCopy 安全处理非 TTY 容器多路复用流，规避包头二进制乱码，彻底净化输出
		_, err := stdcopy.StdCopy(&stdoutBuf, &stderrBuf, resp.Reader)
		doneChan <- err
	}()

	select {
	case err := <-doneChan:
		if err != nil {
			return &ContainerStatusReport{Status: "error", IP: "error", Error: fmt.Sprintf("读取流出错: %v", err)}, nil
		}
	case <-time.After(10 * time.Second):
		return &ContainerStatusReport{Status: "running", IP: "error", Error: "获取出口 IP 动作执行超时"}, nil
	}

	cleanIP := strings.TrimSpace(stdoutBuf.String())
	if ipv4Regexp.MatchString(cleanIP) {
		return &ContainerStatusReport{Status: "running", IP: cleanIP, Error: ""}, nil
	}

	return &ContainerStatusReport{
		Status: "error",
		IP:     "error",
		Error:  fmt.Sprintf("无法获取有效出口 IP, stdout: %s, stderr: %s", cleanIP, strings.TrimSpace(stderrBuf.String())),
	}, nil
}

// GetContainerLogs 获取出口容器最近日志
func GetContainerLogs(name string, tail int) (string, error) {
	cli, err := getDockerClient()
	if err != nil {
		return "", err
	}

	containerName := containerPrefix + name
	ctx := context.Background()

	_, err = cli.ContainerInspect(ctx, containerName)
	if err != nil {
		return "", fmt.Errorf("容器未创建或已被删除: %v", err)
	}

	options := types.ContainerLogsOptions{
		ShowStdout: true,
		ShowStderr: true,
		Timestamps: true,
		Tail:       strconv.Itoa(tail),
	}

	reader, err := cli.ContainerLogs(ctx, containerName, options)
	if err != nil {
		return "", fmt.Errorf("读取日志流失败: %v", err)
	}
	defer reader.Close()

	var stdoutBuf, stderrBuf bytes.Buffer
	// 拆流去除 Docker Stream 包头
	_, err = stdcopy.StdCopy(&stdoutBuf, &stderrBuf, reader)
	if err != nil {
		return "", fmt.Errorf("解析日志包头失败: %v", err)
	}

	// 拼接错误与正常输出返回
	logs := stdoutBuf.String() + stderrBuf.String()
	if logs == "" {
		return "暂无日志", nil
	}
	return logs, nil
}
